lib/patchelf/saver.rb

```diff
@@ -121,31 +121,26 @@ def patch_runpath(sym = :runpath)
     def patch_needed
       original_needs = dynamic.tags_by_type(:needed)
       @set[:needed].uniq!
+
+      original = original_needs.map(&:name)
+      replace = @set[:needed]
+
       # 3 sets:
       # 1. in original and in needs - remain unchanged
       # 2. in original but not in needs - remove
       # 3. not in original and in needs - append
-      original_needs.each do |n|
-        next if @set[:needed].include?(n.name)
-
-        n.header.d_tag = IGNORE # temporarily mark
-      end
-
-      extra = @set[:needed] - original_needs.map(&:name)
-      original_needs.each do |n|
-        break if extra.empty?
-        next if n.header.d_tag != IGNORE
+      append = replace - original
+      remove = original - replace
 
-        n.header.d_tag = ELFTools::Constants::DT_NEEDED
-        reg_str_table(extra.shift) { |idx| n.header.d_val = idx }
-      end
-      return if extra.empty?
-
-      # no spaces, need append
-      extra.each do |name|
+      append.each do |name|
         tag = lazy_dyn(:needed)
         reg_str_table(name) { |idx| tag.d_val = idx }
       end
+
+      remove.each do |name|
+        dyn = original_needs.find { |n| n.name == name }.header
+        dyn.d_tag = IGNORE
+      end
```

Comment:
Uhm why these changes?
My original design is set the "unneeded" tags as IGRNOE, then we can append newly needed tags by occupying IGNOREd tags if any. This could prevent unnecessary tags to be appended to the dynamic table.

Reply:
> My original design is set the "unneeded" tags as IGRNOE, then we can append newly needed tags by occupying IGNOREd tags if any.

ohkay, i just read the comment and missed this aspect.